pkg/internal/apis/certmanager/v1/defaults.go

```diff
@@ -17,9 +17,34 @@ limitations under the License.
 package v1
 
 import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
 )
 
 func addDefaultingFuncs(scheme *runtime.Scheme) error {
 	return RegisterDefaults(scheme)
 }
+
+func SetDefaults_CertificateSpec(o *cmapi.CertificateSpec) {
+	if o == nil {
+		return
+	}
+	setDefaultRenewBefore(o)
+}
+
+func setDefaultRenewBefore(o *cmapi.CertificateSpec) {
+	if o.RenewBefore != nil {
+		return
+	}
+	if o.Duration == nil {
+		o.RenewBefore = &metav1.Duration{Duration: cmapi.DefaultRenewBefore}
+		return
+	}
+	renewBefore := o.Duration.Duration / 3
+	if renewBefore < cmapi.MinimumRenewBefore {
+		renewBefore = cmapi.MinimumRenewBefore
```

Comment:
what is duration is lower than MinimumRenewBefore?

Reply:
I think we need to be really really careful with this sort of stuff. We won't be able to replicate this defaulting into openapi in future with how it currently stands. I'd expect a validation error here instead of defaulting to our minimum if the user specifies a number smaller than we support.

As I've noted before, I'm really hesitant to begin defaulting in our mutating webhook given we are so close to being able to do defaulting properly via the openapi spec (from k8s 1.18 onwards). Seemingly innocuous changes could really bite us in future.

Can I request we hold this PR and have a full discussion of the implications of it, and what our long term strategy is?

Reply:
@munnerz agreed, some background info: we discussed in the standups that there needs to be a defaulting migration plan written up and discussed that describes what is applied where and why and that includes a strategy to move to 1.18 openapi defaulting